stdlib/Artifacts/src/Artifacts.jl

```diff
@@ -0,0 +1,634 @@
+the override: overriding to an on-disk location through an absolutet path, and
+overriding to another artifact by its content-hash.
+"""
+const ARTIFACT_OVERRIDES = Ref{Union{Dict{Symbol,Any},Nothing}}(nothing)
+function load_overrides(;force::Bool = false)
+    if ARTIFACT_OVERRIDES[] !== nothing && !force
+        return ARTIFACT_OVERRIDES[]
+    end
+
+    # We organize our artifact location overrides into two camps:
+    #  - overrides per UUID with artifact names mapped to a new location
+    #  - overrides per hash, mapped to a new location.
+    #
+    # Overrides per UUID/bound name are intercepted upon Artifacts.toml load, and new
+    # entries within the "hash" overrides are generated on-the-fly.  Thus, all redirects
+    # mechanisticly happen through the "hash" overrides.
+    overrides = Dict{Symbol,Any}(
+        # Overrides by UUID
+        :UUID => Dict{Base.UUID,Dict{String,Union{String,SHA1}}}(),
+
+        # Overrides by hash
+        :hash => Dict{SHA1,Union{String,SHA1}}(),
+    )
+
+    for override_file in reverse(artifacts_dirs("Overrides.toml"))
+        !isfile(override_file) && continue
+
+        # Load the toml file
+        depot_override_dict = parse_toml(override_file)
+
+        function parse_mapping(mapping::String, name::String)
+            if !isabspath(mapping) && !isempty(mapping)
+                try
+                    mapping = Base.SHA1(mapping)
+                catch e
+                    @error("Invalid override in '$(override_file)': entry '$(name)' must map to an absolute path or SHA1 hash!")
+                    rethrow()
+                end
+            end
+            return mapping
+        end
+        function parse_mapping(mapping::Dict, name::String)
+            return Dict(k => parse_mapping(v, name) for (k, v) in mapping)
+        end
+
+        for (k, mapping) in depot_override_dict
+            # First, parse the mapping. Is it an absolute path, a valid SHA1-hash, or neither?
+            try
+                mapping = parse_mapping(mapping, k)
+            catch
+                @error("Invalid override in '$(override_file)': failed to parse entry `$(k)`")
+                continue
+            end
+
+            # Next, determine if this is a hash override or a UUID/name override
+            if isa(mapping, String) || isa(mapping, SHA1)
+                # if this mapping is a direct mapping (e.g. a String), store it as a hash override
+                hash = try
+                    Base.SHA1(hex2bytes(k))
+                catch
+                    @error("Invalid override in '$(override_file)': Invalid SHA1 hash '$(k)'")
+                    continue
+                end
+
+                # If this mapping is the empty string, un-override it
+                if mapping == ""
+                    delete!(overrides[:hash], hash)
+                else
+                    overrides[:hash][hash] = mapping
+                end
+            elseif isa(mapping, Dict{String, Any})
```

Comment:
maybe make this into an `else` and type assert or an `else` and error. It just feels a little bit scary that we maybe missed that `mapping` is some other type and that it would then just go through and do nothing?

Reply:
Considering that `mapping` is coming from `parse_mapping`, I think the types are very tightly controlled.  But I believe in myself less than the compiler, so I still think it's a good idea.  :)